### PR TITLE
G3

### PR DESCRIPTION
**Pull request by anyagenza**

### CODE REVIEW

File: Lesson1 (Sequence containers)/TaskFunctions.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#ifndef TASKFUNCTIONS_H
+#define TASKFUNCTIONS_H
+#include <iostream>
+#include <vector>
+#include <forward_list>
+
+std::vector<int> combineVectors(std::vector<int>::const_iterator begin1, std::vector<int>::const_iterator end1,
```
Comment (smay1613) [inline Lesson1 (Sequence containers)/TaskFunctions.h:7]: **why these methods are in the second hometask? please create new branch from the master**

File: Lesson3 (Container adapters)/FunctionsForTest.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#include <iostream>
+#include <stack>
+#include <queue>
+#include <memory>
+#include "FunctionsForTest.h"
+
+bool isExpressionValid (const std::string& expression)
+{
+    std::stack<char> MyStack;
```
Comment (smay1613) [inline Lesson3 (Container adapters)/FunctionsForTest.cpp:9]: _naming_

File: Lesson3 (Container adapters)/FunctionsForTest.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    char popStack;
+    bool result = true;
```
Comment (smay1613) [inline Lesson3 (Container adapters)/FunctionsForTest.cpp:11]: _{}_

File: Lesson3 (Container adapters)/FunctionsForTest.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        else
+        {
+            if(MyStack.empty())
+            {
+                result = false;
+                break;
+            }
+            popStack = MyStack.top();
+            MyStack.pop();
+            if (!(((element == ')')&&(popStack=='('))||
```
Comment (smay1613) [inline Lesson3 (Container adapters)/FunctionsForTest.cpp:27]: _extract to the named function_

File: Lesson3 (Container adapters)/FunctionsForTest.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    if(!MyStack.empty())
+    {
+        result = false;
+    }
+    return result;
+}
+
+void linkSamelevel(std::shared_ptr<node> t)
+{
+    std::queue <std::shared_ptr<node>> q;
```
Comment (smay1613) [inline Lesson3 (Container adapters)/FunctionsForTest.cpp:47]: **Not accepted. It's a copy-paste from https://www.geeksforgeeks.org/connect-nodes-level-level-order-traversal/.**

File: Lesson3 (Container adapters)/FunctionsForTest.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    q.push(t);
+    q.push(nullptr);
+    while (!q.empty())
+    {
+        std::shared_ptr<node> p = q.front();
+        q.pop();
+        if (p != nullptr)
```
Comment (smay1613) [inline Lesson3 (Container adapters)/FunctionsForTest.cpp:54]: **what if we pass nullptr to the method?
exit condition will never be executed.**

File: Lesson3 (Container adapters)/FunctionsForTest.cpp (hunk 5 above)
Comment (smay1613) [reply Lesson3 (Container adapters)/FunctionsForTest.cpp:47]: So, here is another task for you:
Implement binary tree wrapper class.
You should implement add/pop (by key and by iterator)/clear methods.
It must support level order traversal with iterators (so, you must implement your own iterator).
With this class, write 15 unit tests for it (and cover the main public operations).

This task must be done till Thursday 2.00 PM (03 October).

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: Lesson1 (Sequence containers)/TaskFunctions.h (hunk 1 above)
Comment (anyagenza) [reply Lesson1 (Sequence containers)/TaskFunctions.h:7]: where should they be?

Comment (smay1613) [reply Lesson1 (Sequence containers)/TaskFunctions.h:7]: in another branch.
you should create a new branch (branched from master) for each hometask

Comment (anyagenza) [reply Lesson1 (Sequence containers)/TaskFunctions.h:7]: but i have created G3

Comment (smay1613) [reply Lesson1 (Sequence containers)/TaskFunctions.h:7]: looks like you've branched from the last homework. so, changes from last homework are included in homework2 branch.
and when you're comaring it with master, the diff is first homework + second

**Pull request closed without merging**